src/eckit/system/Library.cc

```diff
@@ -109,23 +109,30 @@ std::string Library::libraryPath() const {
     return libraryPath_;
 }
 
+// Store a pointer to the channel, rather than the object itself, as otherwise it can be
+// moved by std::map...
+thread_local std::map<const Library*, std::unique_ptr<Channel>> debugChannels;
+
 Channel& Library::debugChannel() const {
-    AutoLock<Mutex> lock(mutex_);
 
-    if (debugChannel_) {
-        return *debugChannel_;
+    auto it = debugChannels.find(this);
+    if (it != debugChannels.end()) {
+        return *it->second;
     }
 
     std::string s = prefix_ + "_DEBUG";
 
+    std::unique_ptr<Channel> newChannel;
     if (debug_) {
-        debugChannel_.reset(new Channel(new PrefixTarget(s)));
+        newChannel = std::make_unique<Channel>(new PrefixTarget(s));
     }
     else {
-        debugChannel_.reset(new Channel());
+        newChannel = std::make_unique<Channel>();
     }
 
-    return *debugChannel_;
+    Channel& retval(*newChannel);
+    debugChannels.emplace(this, newChannel.release());
```

Comment:
I'd try to avoid using `release()` on unique pointer.
We can directly `emplace` and use the result iterator:

```
return *debugChannels.emplace(this, 
    debug_ ? std::make_unique<Channel>() 
                   : std::make_unique<Channel>(new Prefix Target(s))
).first->get();
```

If you don't like the conditional inside the emplace, I'd put it in a function/lambda. But all in all I think it's better to emplace directly instead of releasing a unique pointer while referring to its value.

Technically also instead of `release()` a `std::move(newChannel)` should also keep the pointer inside alive - but avoiding it at all might be cleaner.

Reply:
I'm making changes now  and hope that we can merge this soon